core/src/main/scala/org/apache/spark/api/r/SerDe.scala

```diff
@@ -46,9 +46,18 @@ private[spark] object SerDe {
     dis.readByte().toChar
   }
 
-  def readObject(dis: DataInputStream): Object = {
+  def readObject(dis: DataInputStream, typeName: String = ""): Object = {
```

Comment:
Yeah I think its fair assumption that the SerDe will return a double in case the schema type is a float. If its not a double it means something went wrong somewhere down the line ? If we want to be really careful we could add a check with `isInstanceOf[Double]` and throw an exception saying `Unexpected type: Expected Double got <>`. 

BTW the reason I'm trying to move this out of SerDe is that the `readObject` code path is used by everything else while the float, double issue only comes up in the case where we create a DataFrame from a local R object.